canonical-web-and-design/maas-ui
# Fetch general data for repository adding/editing

## Done
- Created general action creator
- Fetch general data `components_to_disable`, `known_architectures` and `pockets_to_disable` in RepositoryForm
- Rewrote `general` state structure so that each field manages its own loading and loaded state.
``` js
general: {
  knownArchitectures: {
    loading: false,
    loaded: false,
    data: []
  },
  osInfo: {
    loaded: false,
    loading: false,
    data: {}
  },
...
```
This has ended up creating a loooooot of similar/duplicated code for reducers and selectors, but I think it's easier to reason about when everything is cleanly separated. Further down the line we could figure out the parts to refactor. 

It's also a bit nested which is unfortunate, I guess the alternative would be to have each loading/loaded field at the top level, i.e.: 
``` js
general: {
  knownArchitectures: [],
  knownArchitecturesLoading: false,
  knownArchitecturesLoaded: false,
  osInfo: {},
  osInfoLoading: false,
  osInfoLoaded: false,
...
```
However would we have errors as a separate field as well? Since there are 17 general websocket methods it seems like this general object could get pretty wide (though not sure if that's necessarily a bad thing).

## QA
- Go to /repositories and click on Add PPA, Add repository or edit a repo
- Check that redux actions are dispatched to fetch general data and it is reflected in state
- Refresh the page and check that a loader shows while the repos and general data are fetched
